File: 0383-ransom-note/0383-ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        int n = ransomNote.size();
        int m = magazine.size();

        if(n > m){
            return false;
        }
        //initialize an array of alphabets counter of size 26
        vector<int> alphabet_counter(26,0);
        for(int i = 0; i < m; i++){
            //increment the counter for that character
            alphabet_counter[magazine[i]-'a']++;
        }

        for(int i = 0; i<n;i++){
            // if the charachter doesn't exists in magazine for ransomNote, we return false
            if(alphabet_counter[ransomNote[i]-'a'] == 0){
                return false;
            }
            //if it does exists, we decrement the counter of that charachter from its own index
            alphabet_counter[ransomNote[i]-'a']--;
        }
        return true;       
    }
};

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
		if (ransomNote.length() > magazine.length()) return false;
        int[] alphabets_counter = new int[26];
        
        for (char c : magazine.toCharArray())
            alphabets_counter[c-'a']++;

        for (char c : ransomNote.toCharArray()){
            if (alphabets_counter[c-'a'] == 0) return false;
            alphabets_counter[c-'a']--;
        }
        return true;
    }
}

/*
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int n = magazine.length();
        int m = ransomNote.length();

        Map<Character, Integer> mp = new HashMap<>();

        for(char ch : magazine.toCharArray()){
            mp.put(ch, mp.getOrDefault(ch,0)+1);
        }

        for(char ch : ransomNote.toCharArray()){
            if(!mp.containsKey(ch) || mp.get(ch) == 0){
                return false;
            }
            mp.put(ch, mp.get(ch)-1);
            
        }

        return true;
    }
}
*/


/*
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int n = magazine.length();
        int m = ransomNote.length();

        Map<Character, Integer> mp = new HashMap<>();

        for(char ch : magazine.toCharArray()){
            mp.put(ch, mp.getOrDefault(ch,0)+1);
        }

        for(char ch : ransomNote.toCharArray()){
            if(mp.containsKey(ch)){
                mp.put(ch, mp.get(ch)-1);
            }
            return false;
        }

        return true;
    }
}
*/